File: gcs/ui/theme.py
"""Centralised colours and a dark stylesheet (one place to restyle the app).

Palette echoes the firmware's GitHub-dark UI plus a Mission-Planner-style HUD
(blue sky / brown ground / white lines / amber aircraft symbol). The QSS aims
for a modern, glanceable fly-view: soft elevated cards, rounded controls, a
clear accent, and thin scrollbars — readable from a wide desktop down to a small
high-DPI panel.
"""
from __future__ import annotations

from PySide6.QtGui import QColor

from ..domain.telemetry import Severity

# ── general UI ──────────────────────────────────────────────────────────────
BG = "#0a0e14"            # app backdrop (a touch deeper than the cards)
PANEL = "#141a22"         # card surface
PANEL_HI = "#1b232d"      # raised / hover surface
PANEL_BORDER = "#2a323d"  # hairline border
TEXT = "#e6edf3"
TEXT_DIM = "#8b97a7"
ACCENT = "#ffb000"        # amber — aircraft symbol, active highlights
ACCENT2 = "#4c9bff"       # blue — interactive accents (focus, links)
GOOD = "#3fb950"          # green — connected / disarmed-safe
BAD = "#ff5d52"           # red — error / armed / disconnected
WARN = "#e3a008"          # amber — warnings

# ── HUD (modern PFD/EFIS palette) ────────────────────────────────────────────
# Sky and ground are drawn as vertical gradients (deep → bright at the horizon).
HUD_SKY_TOP = QColor("#0b3a66")
HUD_SKY_HORIZON = QColor("#2f8fd6")
HUD_GROUND_HORIZON = QColor("#a06a30")
HUD_GROUND_BOTTOM = QColor("#3c2913")
HUD_LINE = QColor("#eef4f9")          # near-white horizon / ladder lines
HUD_ACCENT = QColor(ACCENT)            # amber aircraft boresight
HUD_CYAN = QColor("#39e1ff")           # cyan bug / selected pointer
HUD_TAPE_BG = QColor(8, 12, 18, 175)
HUD_TAPE_TICK = QColor("#cdd9e3")
HUD_BOX_BG = QColor(8, 12, 18, 225)
HUD_CHIP_BG = QColor(8, 12, 18, 160)
HUD_TEXT = QColor("#ffffff")
HUD_TEXT_DIM = QColor("#9fb1bf")

# ── severity → colour for the message log ───────────────────────────────────
_SEVERITY_COLOR = {
    Severity.EMERGENCY: BAD,
    Severity.ALERT: BAD,
    Severity.CRITICAL: BAD,
    Severity.ERROR: BAD,
    Severity.WARNING: WARN,
    Severity.NOTICE: "#58a6ff",
    Severity.INFO: TEXT,
    Severity.DEBUG: TEXT_DIM,
}


def severity_color(sev: Severity) -> str:
    return _SEVERITY_COLOR.get(sev, TEXT)


STYLESHEET = f"""
* {{
    outline: none;
}}
QWidget {{
    background-color: {BG};
    color: {TEXT};
    font-family: "Segoe UI Variable", "Segoe UI", "Inter", "DejaVu Sans", sans-serif;
    font-size: 13px;
}}
QToolTip {{
    background-color: {PANEL_HI};
    color: {TEXT};
    border: 1px solid {PANEL_BORDER};
    border-radius: 6px;
    padding: 5px 8px;
}}

/* ── cards ──────────────────────────────────────────────────────────────── */
QFrame#Panel {{
    background-color: rgba(20, 26, 34, 0.94);
    border: 1px solid {PANEL_BORDER};
    border-radius: 12px;
}}
QLabel#PanelTitle {{
    color: {TEXT_DIM};
    font-size: 11px;
    font-weight: 700;
    letter-spacing: 1.5px;
}}

/* ── buttons ────────────────────────────────────────────────────────────── */
QPushButton {{
    background-color: {PANEL_HI};
    border: 1px solid {PANEL_BORDER};
    border-radius: 8px;
    padding: 8px 12px;
    font-weight: 600;
}}
QPushButton:hover {{ background-color: #232c37; border-color: #3a4654; }}
QPushButton:pressed {{ background-color: #161d25; }}
QPushButton:disabled {{ color: #4b5360; background-color: #131820; border-color: #20272f; }}
QPushButton:checked {{
    background-color: #243140;
    border-color: {ACCENT2};
    color: {TEXT};
}}

QPushButton#Arm {{ background-color: {GOOD}; color: #062a10; border: none; }}
QPushButton#Arm:hover {{ background-color: #4fd463; }}
QPushButton#Disarm {{ background-color: {BAD}; color: #2a0707; border: none; }}
QPushButton#Disarm:hover {{ background-color: #ff7068; }}
QPushButton#Mode {{ font-size: 15px; padding: 12px 8px; }}
QPushButton#Connect {{ background-color: #238636; color: #ffffff; border: none; padding: 8px 18px; }}
QPushButton#Connect:hover {{ background-color: #2ea043; }}

/* small square icon button (refresh, zoom, swap…) */
QPushButton#IconButton {{
    padding: 6px;
    min-width: 30px;
    border-radius: 8px;
    font-size: 15px;
    font-weight: 700;
}}
/* quiet "ghost" button used on chips/overlays */
QPushButton#Ghost {{
    background-color: rgba(255, 255, 255, 0.06);
    border: 1px solid rgba(255, 255, 255, 0.10);
    border-radius: 8px;
    padding: 5px 12px;
}}
QPushButton#Ghost:hover {{ background-color: rgba(255, 255, 255, 0.12); }}
QPushButton#ChipDisconnect {{
    background-color: rgba(255, 93, 82, 0.16);
    border: 1px solid rgba(255, 93, 82, 0.45);
    color: #ff8a82;
    border-radius: 8px;
    padding: 5px 12px;
}}
QPushButton#ChipDisconnect:hover {{ background-color: rgba(255, 93, 82, 0.28); }}

/* ── inputs ─────────────────────────────────────────────────────────────── */
QComboBox, QSpinBox, QLineEdit {{
    background-color: #0c1118;
    border: 1px solid {PANEL_BORDER};
    border-radius: 8px;
    padding: 6px 9px;
    selection-background-color: {ACCENT2};
}}
QComboBox:hover, QSpinBox:hover, QLineEdit:hover {{ border-color: #3a4654; }}
QComboBox:focus, QSpinBox:focus, QLineEdit:focus {{ border-color: {ACCENT2}; }}
QComboBox::drop-down {{ border: none; width: 20px; }}
QComboBox::down-arrow {{
    image: none;
    border-left: 4px solid transparent;
    border-right: 4px solid transparent;
    border-top: 5px solid {TEXT_DIM};
    margin-right: 8px;
}}
QComboBox QAbstractItemView {{
    background-color: {PANEL};
    border: 1px solid {PANEL_BORDER};
    border-radius: 8px;
    selection-background-color: {PANEL_HI};
    outline: none;
    padding: 4px;
}}

/* ── check box ──────────────────────────────────────────────────────────── */
QCheckBox {{ spacing: 7px; color: {TEXT_DIM}; }}
QCheckBox::indicator {{
    width: 16px; height: 16px;
    border: 1px solid {PANEL_BORDER};
    border-radius: 5px;
    background-color: #0c1118;
}}
QCheckBox::indicator:hover {{ border-color: {ACCENT2}; }}
QCheckBox::indicator:checked {{
    background-color: {ACCENT2};
    border-color: {ACCENT2};
}}

/* ── text panels ────────────────────────────────────────────────────────── */
QPlainTextEdit, QTextEdit {{
    background-color: #080b10;
    border: 1px solid {PANEL_BORDER};
    border-radius: 8px;
    font-family: "Cascadia Mono", "Consolas", "DejaVu Sans Mono", monospace;
    font-size: 12px;
}}

/* ── thin scrollbars ────────────────────────────────────────────────────── */
QScrollBar:vertical {{ background: transparent; width: 10px; margin: 2px; }}
QScrollBar::handle:vertical {{
    background: #313b47; border-radius: 5px; min-height: 28px;
}}
QScrollBar::handle:vertical:hover {{ background: #3e4a59; }}
QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical {{ height: 0; }}
QScrollBar::add-page:vertical, QScrollBar::sub-page:vertical {{ background: transparent; }}
QScrollBar:horizontal {{ background: transparent; height: 10px; margin: 2px; }}
QScrollBar::handle:horizontal {{
    background: #313b47; border-radius: 5px; min-width: 28px;
}}
QScrollBar::handle:horizontal:hover {{ background: #3e4a59; }}
QScrollBar::add-line:horizontal, QScrollBar::sub-line:horizontal {{ width: 0; }}
QScrollBar::add-page:horizontal, QScrollBar::sub-page:horizontal {{ background: transparent; }}

/* ── connection status chip (shown once a link is up) ───────────────────── */
QFrame#ConnChip {{
    background-color: rgba(20, 26, 34, 0.92);
    border: 1px solid {PANEL_BORDER};
    border-radius: 14px;
}}
QLabel#ConnLabel {{ font-weight: 600; color: {TEXT}; }}

/* ── combined top pill (connection status + active view's controls) ─────── */
QFrame#TopBar {{
    background-color: rgba(20, 26, 34, 0.92);
    border: 1px solid {PANEL_BORDER};
    border-radius: 12px;
}}
QFrame#TopBar QLabel {{ color: {TEXT_DIM}; }}
QFrame#TopBar QLabel#ConnLabel {{ color: {TEXT}; font-weight: 600; }}
QFrame#TopBar QPushButton {{ padding: 5px 10px; }}
QFrame#TopBar QPushButton#IconButton {{ min-width: 26px; padding: 4px; font-size: 14px; }}
QFrame#TopBar QComboBox {{ padding: 4px 8px; }}

/* ── camera placeholder (font size is set dynamically to match the tile) ── */
QLabel#CamPlaceholder {{ color: {TEXT_DIM}; }}
"""

File: cpp/src/ui/theme.cpp
#include "ui/theme.h"

#include <QPainter>

namespace gcs::theme {

QPixmap tintedPixmap(const QString &resourcePath, const QColor &color)
{
    QPixmap src(resourcePath);
    if (src.isNull())
        return src;
    QPixmap out(src.size());
    out.fill(Qt::transparent);
    QPainter p(&out);
    p.drawPixmap(0, 0, src);                                  // hình dạng (alpha)
    p.setCompositionMode(QPainter::CompositionMode_SourceIn); // giữ alpha, thay màu
    p.fillRect(out.rect(), color);
    p.end();
    out.setDevicePixelRatio(src.devicePixelRatio());
    return out;
}

QIcon railIcon(const QString &resourcePath)
{
    const QPixmap src(resourcePath);
    if (src.isNull())
        return QIcon();
    QIcon icon;
    // thường (chưa chọn): mực sáng
    icon.addPixmap(tintedPixmap(resourcePath, QColor(TEXT)), QIcon::Normal, QIcon::Off);
    // được chọn: nền xanh sáng nên dùng mực tối cho tương phản
    const QPixmap dark = tintedPixmap(resourcePath, QColor("#04101f"));
    icon.addPixmap(dark, QIcon::Normal, QIcon::On);
    icon.addPixmap(dark, QIcon::Active, QIcon::On);
    // vô hiệu: xám mờ như QSS
    icon.addPixmap(tintedPixmap(resourcePath, QColor("#5a6472")), QIcon::Disabled, QIcon::Off);
    return icon;
}

QString severityColor(domain::Severity sev)
{
    using domain::Severity;
    switch (sev) {
    case Severity::Emergency:
    case Severity::Alert:
    case Severity::Critical:
    case Severity::Error:   return QString::fromLatin1(BAD);
    case Severity::Warning: return QString::fromLatin1(WARN);
    case Severity::Notice:  return QStringLiteral("#58a6ff");
    case Severity::Info:    return QString::fromLatin1(TEXT);
    case Severity::Debug:   return QString::fromLatin1(TEXT_DIM);
    }
    return QString::fromLatin1(TEXT);
}

QString stylesheet()
{
    // Cổng chuyển thẳng từ theme.py; giá trị màu đã nội tuyến.
    return QStringLiteral(R"QSS(
* { outline: none; }
QWidget {
    background-color: #0a0e14;
    color: #eef2f7;
    font-family: "Segoe UI Variable", "Segoe UI", "Inter", "DejaVu Sans", sans-serif;
    font-size: 13px;
}
QToolTip {
    background-color: #1b232d;
    color: #eef2f7;
    border: 1px solid #3a4654;
    border-radius: 6px;
    padding: 5px 8px;
}

/* cards */
QFrame#Panel { background-color: transparent; border: none; }
QLabel#PanelTitle {
    color: #b8c3d1;
    font-size: 11px;
    font-weight: 800;
    letter-spacing: 1.6px;
}

/* buttons */
QPushButton {
    background-color: #1b232d;
    border: 1px solid #313b47;
    border-radius: 8px;
    padding: 9px 14px;
    font-weight: 600;
    color: #eef2f7;
}
QPushButton:hover { background-color: #232c37; border-color: #4a586a; }
QPushButton:pressed { background-color: #161d25; }
QPushButton:disabled { color: #5a6472; background-color: #131820; border-color: #20272f; }
QPushButton:checked { background-color: #223449; border-color: #5aa6ff; color: #ffffff; }

QPushButton#Arm { background-color: #4cc463; color: #032010; border: none; }
QPushButton#Arm:hover { background-color: #5cd873; }
QPushButton#Disarm { background-color: #ff6459; color: #2a0707; border: none; }
QPushButton#Disarm:hover { background-color: #ff7a70; }
QPushButton#Mode { font-size: 15px; padding: 12px 8px; }
QPushButton#Connect { background-color: #2ea043; color: #ffffff; border: none; padding: 8px 18px; font-weight: 700; }
QPushButton#Connect:hover { background-color: #3ab953; }
QPushButton#Connect:pressed { background-color: #268038; }

QPushButton#IconButton {
    padding: 6px; min-width: 36px; min-height: 36px;
    border-radius: 8px; font-size: 16px; font-weight: 700;
}
QPushButton#Ghost {
    background-color: rgba(255, 255, 255, 0.08);
    border: 1px solid rgba(255, 255, 255, 0.16);
    border-radius: 8px; padding: 5px 12px; color: #eef2f7;
}
QPushButton#Ghost:hover { background-color: rgba(255, 255, 255, 0.15); border-color: rgba(255, 255, 255, 0.24); }
QPushButton#ChipDisconnect {
    background-color: rgba(255, 93, 82, 0.16);
    border: 1px solid rgba(255, 93, 82, 0.45);
    color: #ff8a82; border-radius: 8px; padding: 5px 12px;
}
QPushButton#ChipDisconnect:hover { background-color: rgba(255, 93, 82, 0.28); }

/* inputs */
QComboBox, QSpinBox, QLineEdit, QDoubleSpinBox {
    background-color: #0c1118;
    border: 1px solid #333d4a;
    border-radius: 8px;
    padding: 6px 9px;
    color: #eef2f7;
    selection-background-color: #5aa6ff;
    selection-color: #061424;
}
QComboBox:hover, QSpinBox:hover, QLineEdit:hover, QDoubleSpinBox:hover { border-color: #465566; }
QComboBox:focus, QSpinBox:focus, QLineEdit:focus, QDoubleSpinBox:focus {
    border: 2px solid #5aa6ff; padding: 5px 8px;
}
QComboBox::drop-down { border: none; width: 20px; }
QComboBox::down-arrow {
    image: none;
    border-left: 4px solid transparent;
    border-right: 4px solid transparent;
    border-top: 5px solid #aab6c4;
    margin-right: 8px;
}
QComboBox::down-arrow:hover { border-top-color: #eef2f7; }
QComboBox QAbstractItemView {
    background-color: #141a22;
    border: 1px solid #333d4a;
    border-radius: 8px;
    selection-background-color: #223449;
    selection-color: #ffffff;
    outline: none; padding: 4px;
}

/* check box */
QCheckBox { spacing: 7px; color: #cdd6e0; }
QCheckBox::indicator {
    width: 16px; height: 16px;
    border: 1px solid #3a4654; border-radius: 5px; background-color: #0c1118;
}
QCheckBox::indicator:hover { border-color: #5aa6ff; }
QCheckBox::indicator:checked { background-color: #5aa6ff; border-color: #5aa6ff; }

/* text panels */
QPlainTextEdit, QTextEdit {
    background-color: #080b10;
    border: 1px solid #2a323d;
    border-radius: 8px;
    font-family: "Cascadia Mono", "Consolas", "DejaVu Sans Mono", monospace;
    font-size: 12px;
}

/* scrollbars */
QScrollBar:vertical { background: transparent; width: 10px; margin: 2px; }
QScrollBar::handle:vertical { background: #313b47; border-radius: 5px; min-height: 28px; }
QScrollBar::handle:vertical:hover { background: #3e4a59; }
QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical { height: 0; }
QScrollBar::add-page:vertical, QScrollBar::sub-page:vertical { background: transparent; }
QScrollBar:horizontal { background: transparent; height: 10px; margin: 2px; }
QScrollBar::handle:horizontal { background: #313b47; border-radius: 5px; min-width: 28px; }
QScrollBar::handle:horizontal:hover { background: #3e4a59; }
QScrollBar::add-line:horizontal, QScrollBar::sub-line:horizontal { width: 0; }
QScrollBar::add-page:horizontal, QScrollBar::sub-page:horizontal { background: transparent; }

/* connection chip */
QFrame#ConnChip {
    background-color: rgba(20, 26, 34, 0.92);
    border: 1px solid #3a4654; border-radius: 14px;
}
QLabel#ConnLabel { font-weight: 600; color: #eef2f7; }

/* thẻ kết nối (acrylic) — thanh này tự vẽ nền kính mờ trong paintEvent, nên QSS
   phải giữ trong suốt; và các nhãn con phải trong suốt để không hiện hộp đen đục
   do luật "QWidget { background-color }" ở trên. */
QFrame#ConnBar { background-color: transparent; border: none; }
QFrame#ConnBar QLabel { background: transparent; color: #aab6c4; }
/* các container (QStackedWidget + trang serial/udp/tcp) cũng phải trong suốt,
   nếu không chúng tạo một dải tối sau nhóm Cổng/Baud. Chỉ nhắm container trực
   tiếp (> QWidget) để KHÔNG chạm nền tối cố ý của các ô combo/spinbox/lineedit. */
QFrame#ConnBar QStackedWidget { background: transparent; }
QFrame#ConnBar QStackedWidget > QWidget { background: transparent; }

/* control dock (acrylic) */
QFrame#TopBar { background-color: transparent; border: none; }
QFrame#TopBar QLabel { color: #aab6c4; }
QFrame#TopBar QLabel#ConnLabel { color: #eef2f7; font-weight: 600; }
QFrame#TopBar QPushButton { padding: 6px 11px; }
QFrame#TopBar QPushButton#IconButton { min-width: 30px; min-height: 30px; padding: 4px; font-size: 15px; }
QFrame#TopBar QComboBox { padding: 5px 8px; }

QLabel#CamPlaceholder { color: #aab6c4; }

/* top status bar */
QFrame#StatusBar { background-color: transparent; border: none; }
QFrame#StatusBar QLabel { color: #eef2f7; }
QLabel#Vehicle, QLabel#ViewTag, QLabel#ConnDot { background: transparent; }
QLabel#Vehicle { font-size: 15px; font-weight: 700; }
QLabel#ViewTag { color: #aab6c4; font-weight: 700; letter-spacing: 0.6px; }
QLabel#StatChip {
    background-color: rgba(255, 255, 255, 0.08);
    border: 1px solid rgba(255, 255, 255, 0.14);
    border-radius: 9px; padding: 4px 10px;
    color: #eef2f7;
    font-family: "Cascadia Mono", "Consolas", monospace; font-weight: 700;
}
QPushButton#ModePill {
    background-color: rgba(90, 166, 255, 0.16);
    border: 1px solid rgba(90, 166, 255, 0.45);
    color: #cfe2ff;
    border-radius: 14px; padding: 6px 16px; font-weight: 800;
}
QPushButton#ModePill:hover { background-color: rgba(90, 166, 255, 0.28); border-color: rgba(90, 166, 255, 0.65); }
QPushButton#ArmPill {
    background-color: rgba(63, 185, 80, 0.18);
    border: 1px solid rgba(63, 185, 80, 0.55);
    color: #7ee48a; border-radius: 14px; padding: 6px 16px; font-weight: 800;
}
QPushButton#ArmPill:hover { background-color: rgba(63, 185, 80, 0.30); }
QPushButton#DisarmPill {
    background-color: rgba(255, 93, 82, 0.20);
    border: 1px solid rgba(255, 93, 82, 0.60);
    color: #ff9089; border-radius: 14px; padding: 6px 16px; font-weight: 800;
}
QPushButton#DisarmPill:hover { background-color: rgba(255, 93, 82, 0.32); }
QPushButton#ArmPill:disabled, QPushButton#DisarmPill:disabled, QPushButton#ModePill:disabled {
    color: #555e6b; background-color: rgba(255,255,255,0.03); border-color: rgba(255,255,255,0.06);
}

/* left mode/action rail */
QFrame#ModeRail { background-color: transparent; border: none; }
QToolButton#RailBtn {
    background-color: rgba(20, 26, 34, 0.62);
    border: 1px solid rgba(255, 255, 255, 0.12);
    border-radius: 12px; padding: 6px 2px; color: #eef2f7;
    font-size: 10px; font-weight: 700; letter-spacing: 0.5px;
}
QToolButton#RailBtn:hover { background-color: rgba(36, 44, 56, 0.82); border-color: rgba(255,255,255,0.22); }
QToolButton#RailBtn:checked { background-color: #5aa6ff; border-color: #5aa6ff; color: #04101f; }
QToolButton#RailBtn:disabled { color: #5a6472; background-color: rgba(20,26,34,0.4); }
)QSS");
}

} // namespace gcs::theme
